examples/datagen_with_al/run.py

```diff
@@ -0,0 +1,867 @@
+    )
+    # logger.warning(
+    #     "Process rank: %s, device: %s, n_gpu: %s, distributed training: %s, 16-bits training: %s",
+    #     training_args.local_rank,
+    #     training_args.device,
+    #     training_args.n_gpu,
+    #     bool(training_args.local_rank != -1),
+    #     training_args.fp16,
+    # )
+    logger.info("RC Training parameters: %s", rc_training_args.to_dict())
+    logger.info("QG Training parameters: %s", qg_training_args.to_dict())
+
+    # Set seed
+    set_seed(rc_training_args.seed)
+    set_seed(qg_training_args.seed)
+
+    # load datasets
+    if al_args.do_al or rc_training_args.do_train or qg_training_args.do_train:
+        train_dataset = get_datasets(data_args.train_dataset, concatenate=True)
+        train_dataset = expand_answers(train_dataset, separate_answers=False)
+    if rc_training_args.do_eval or qg_training_args.do_eval:
+        validation_dataset = get_datasets(data_args.eval_dataset)
+        validation_dataset = expand_answers(validation_dataset, separate_answers=False)
+
+    def preprocess_fn_wrapper(
+        training_args: TrainingArguments,
+        process_fn: Callable,
+        add_examples_to_output: bool = False,
+    ) -> Callable[[Dataset], Tuple[Dataset, Dataset]]:
+        # wrapper to have access to training_args
+        def wrapped_process_fn(examples) -> Tuple[Dataset, Dataset]:
+            with training_args.main_process_first(desc="Dataset pre-processing"):
+                result = process_fn(examples)
+            if add_examples_to_output:
+                # the pre-processing function returns only processed examples but we need the examples as well
+                return examples, result
+            # otherwise return the result, should be examples and processed examples
+            return result
+
+        return wrapped_process_fn
+
+    ## RC setup
+
+    if rc_model_args.model_name_or_path is not None:
+        # detecting last checkpoint
+        last_checkpoint = None
+        if (
+            os.path.isdir(rc_training_args.output_dir)
+            and rc_training_args.do_train
+            and not rc_training_args.overwrite_output_dir
+        ):
+            last_checkpoint = get_last_checkpoint(rc_training_args.output_dir)
+            if last_checkpoint is None and len(os.listdir(rc_training_args.output_dir)) > 0:
+                raise ValueError(
+                    f"Output directory ({rc_training_args.output_dir}) already exists and is not empty. "
+                    "Use --rc_overwrite_output_dir to overcome."
+                )
+            elif last_checkpoint is not None and rc_training_args.resume_from_checkpoint is None:
+                logger.info(
+                    f"Checkpoint detected, resuming training at {last_checkpoint}. To avoid this behavior, change "
+                    "the `--rc_output_dir` or add `--rc_overwrite_output_dir` to train from scratch."
+                )
+
+        task_heads = EXTRACTIVE_HEAD
```

Comment:
Can we not extend/use existing run_mrc.py to do the same thing? I see many repetitive code snippets here and there.